nengo_gui/static/datastore.js

```diff
@@ -146,3 +148,210 @@ Nengo.DataStore.prototype.get_last_data = function() {
+ * the storage limit set by the Nengo.SimControl.
+ */
+Nengo.GrowableDataStore.prototype.update = function() {
+    /** figure out how many extra values we have (values whose time stamp is
+     * outside the range to keep)
+     */
+    var offset = this.get_offset();
+    var extra = 0;
+    var limit = this.sim.time_slider.last_time -
+                this.sim.time_slider.kept_time;
+    while (this.times[extra] < limit) {
+        extra += 1;
+    }
+
+    /** remove the extra data */
+    if (extra > 0) {
+        this.times = this.times.slice(extra);
+        for (var i = 0; i < this.data.length; i++) {
+            if(extra - offset[i] >= 0){
+                this.data[i] = this.data[i].slice(extra - offset[i]);
+            }
+        }
+    }
+}
+
+/**
+ * Return just the data that is to be shown
+ */
+Nengo.GrowableDataStore.prototype.get_shown_data = function() {
+    var offset = this.get_offset();
+    /* determine time range */
+    var t1 = this.sim.time_slider.first_shown_time;
+    var t2 = t1 + this.sim.time_slider.shown_time;
+
+    /* find the corresponding index values */
+    var index = 0;
+    while (this.times[index] < t1) {
+        index += 1;
+    }
+    // logically, you should start the search for the 
+    var last_index = index;
+    while (this.times[last_index] < t2 && last_index < this.times.length) {
+        last_index += 1;
+    }
+    this.first_shown_index = index;
+
+    /** return the visible slice of the data */
+    var shown = [];
+    var nan_number = 0;
+    var slice_start = 0;
+    for (var i = 0; i < this._dims; i++) {
+
+        if(last_index > offset[i] && offset[i] !== 0){
+
+            if(index < offset[i]){
+                nan_number = offset[i] - index;
+                slice_start = 0;
+            } else {
+                nan_number = 0;
+                slice_start = index - offset[i];
+            }
+
+            shown.push(
+                Array.apply(null, Array(nan_number)).map(function(){return "NaN"}).concat(
+                    this.data[i].slice(slice_start, last_index - offset[i])
+                )
+            );
+
+        } else {
+
+            shown.push(this.data[i].slice(index, last_index));
+
+        }
+    }
+
+    return shown;
+}
+
+Nengo.GrowableDataStore.prototype.get_last_data = function() {
+    var offset = this.get_offset();
+    /* determine time range */
+    var t1 = this.sim.time_slider.first_shown_time;
+    var t2 = t1 + this.sim.time_slider.shown_time;
+
+    /* find the corresponding index values */
+    var last_index = 0;
+    while (this.times[last_index] < t2 && last_index < this.times.length - 1) {
+        last_index += 1;
+    }
+
+    /** return the visible slice of the data */
+    var shown = [];
+    for (var i = 0; i < this._dims; i++) {
+        if(last_index - offset[i] >= 0){
+            shown.push(this.data[i][last_index - offset[i]]);
+        }
+    }
+    return shown;
+}
```

Comment:
Right, I made a separate issue #556 for that, since the optimisation of that code is kind of outside the scope of this pull request. In this pull request I just needed my own jagged datastore.